### PR TITLE
Added support for type based parameter binding (#35496)

### PR DESCRIPTION
**Pull request by davidfowl**
## Customer Impact

We've had several requests to enable MVC style model binding for minimal APIs, for both framework authors and customers who have written basic model binders for various types. We're enabling typed based modeling binding to unlock these scenarios so we can get feedback for [.NET 7.0 for the overall design](https://github.com/dotnet/aspnetcore/issues/35489)

## Testing

Unit testing, lots of it.

## Risk

Low

### CODE REVIEW

**Review by halter73 — APPROVED** [review]

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -739,6 +813,42 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
...
         }
 
+        private static Expression BindParameterFromBindAsync(ParameterInfo parameter, FactoryContext factoryContext)
+        {
+            // We reference the boundValues array by parameter index here
+            var nullability = NullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            // Get the BindAsync method
+            var body = TryParseMethodCache.FindBindAsyncMethod(parameter.ParameterType)!;
```
Comment (pranavkm) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:823]: Super small nit, but the name is no longer an accurate reflection of what's being cached. `BinderMethodCache` or something like that?

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 2]
```diff
@@ -499,6 +499,53 @@ public static bool TryParse(string? value, out MyTryParsableRecord? result)
             }
         }
 
+        private class MyBindAsyncTypeThatThrows
+        {
+            public static ValueTask<object?> BindAsync(HttpContext context)
```
Comment (pranavkm) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:504]: What's the behavior like if the signature does not match? e.g. `ValueTask BindAsync(HttpContext context)`?

**Review by pranavkm — COMMENTED** [review]
I'll trust @halter73 to review the changes to the factory.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:823]: In .NET 7. This shared source is painful.

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs (hunk 2 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:504]: It won't find it

**Comment by davidfowl** [comment]
cc @wtgodbe

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs (hunk 2 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:504]: I could fail if it found the method but had the wrong return type?

Comment (pranavkm) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:504]: Toss it in the Analyzer pile?

Comment (halter73) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:504]: Can we just look for `ValueTask<{type}>` in `FindBindAsyncMethod`? What's the benefit of allowing any object to be returned? I get that we're going to box anyway for structs, but might as well lean on the type-safety checks of the compiler.

We could even throw at startup if a type is used with a `BindAsync(HttpContext)` method with the wrong return type.

**Pull request merged**